lib/experimental/html/class-wp-html-processor.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+/**
+ * @TODO: Handle self-closing foreign elements.
+ * @TODO: Detect non-normative HTML input.
+ * @TODO: Consider parsing non-normative HTML input, support adoption agency algorithm.
+ *
+ * If we support non-normative HTML we can probably handle significantly more
+ * HTML without introducing unexpected results, but I'm not sure yet if we can
+ * handle HTML the same way as the browser, because the section in HTML5 spec
+ * dealing with errors is itself "non-normative" and only issues a few examles.
+ *
+ * Not yet clear is if browsers are full of special one-off cases for "invalid"
+ * input. E.g. it's clear to me how to handle `</b></b></b>` but not clear to how
+ * handle `</p></p></p>` given that `<b>` is a formatting element but `<p>` is
+ * not, that `<p>` itself is a special element.
+ */
+
+class WP_HTML_Processor extends WP_HTML_Tag_Processor {
```

Comment:
This also works with tags, it's even in the method names – I'd be slightly confused to see two classes with such similar names doing similar things. Should we bring these methods to the parent class?

Reply:
it's not unsafe right now because the functions have been renamed to imply they are making assumptions (and those names might change)

but there's still a difference between this and the tag processor that I wrestle with. this is "safe" in that it will preserve the HTML syntax by relying on the tag processor. this is "unsafe" in that it allows one to break the semantic of the HTML by performing operations that would require multiple steps in a DOM, such as removing a range of HTML tags or tag openers or tag closers. it's also "unsafe" in that even though it preserves the syntax, it may not do what you expect it to, even in the renamed functions.


```html
<p>Before<p>Inside</p>
```

For this input there's nothing that will trap the first paragraph and any balanced tag operation will fail after reaching the end of input. Hopefully that will change in time, and we'd properly find `<p>Before` as one outer content and `<p>Inside</p>` as another, but even then our performance domain has shifted greatly from the tag processor.

I'm thinking that as this class develops people may not generally call the tag processor and it will remain as a low-level library. That's why I renamed it to `WP_HTML_Processor`, because I think this could end up as the "entry point" interface for working with HTML. Sistering this is the as-of-yet-still-un-renamed-to `WP_PHPQ` (for lack of a better name) which might go inside here or still might live separately as a class for turning a CSS selector into an HTML query. The HTML class could expose that query instead of our `next_tag()` interface.

```php
$p = new WP_HTML_Processor( $html );
while ( $p->find( 'section + h2 > a[href^="#"]' ) ) {
	$title = $p->get_inner_text();
}
```

It's still not a DOM, we can't call `->parent_node()` or walk backwards (even though we can manually do this with bookmarking), but we can jump to specific places in the document, mangle the attributes, and mangle the content inside it.

Reply:
> it's also "unsafe" in that even though it preserves the syntax, it may not do what you expect it to, even in the renamed functions.

And then it allows you to find an element by a class name, even if it's `table`, and then insert a `div` inside it – which the browsers can handle, but is likely not what you want.

> That's why I renamed it to WP_HTML_Processor, because I think this could end up as the "entry point" interface for working with HTML.

Oh I see, with this insight I'd probably rename the Tag_Processor to something else just eo avoid the confusion – but at this point I don't think it's worth it. 🤷  

> The HTML class could expose that query instead of our next_tag() interface.

I like that!

Reply:
> I'd probably rename the Tag_Processor

I don't know about this. I think based on how careful we were with naming that it will retain the meaning we gave it.

What do you find contradictory about that class, if we introduce an HTML processor?

Reply:
It's not as much that I see a contradiction, as that I worry about confusing everyone with two very similarly named classes (in a similar way as folks get confused about slice vs splice in JS)

Reply:
let's keep thinking about it then. I want to keep the classes separate because of the different performance implications, but I don't know how to address what you're suggesting (apart from `Unsafe` which I'm currently skeptical of using)